src/image-tvg.c

```diff
@@ -0,0 +1,1449 @@
+    if (2 > ctx->in((uint8_t *) data, 2, ctx->in_state)) {
+        return TVG_E_IO_ERROR;
+    }
+    if (data[0] != 0x72 || data[1] != 0x56) {
+        return TVG_E_INVALID_FORMAT;
+    }
+    /* read the version */
+    if (1 > ctx->in(data, 1, ctx->in_state)) {
+        return TVG_E_IO_ERROR;
+    }
+    /* we only support version 1 */
+    if (data[0] != 1) {
+        return TVG_E_NOT_SUPPORTED;
+    }
+    /* read the scale, encoding, and coordinate range: */
+    if (1 > ctx->in(data, 1, ctx->in_state)) {
+        return TVG_E_IO_ERROR;
+    }
+    /* it's all packed in a byte, so crack it up */
+    ctx->scale = TVG_HEADER_DATA_SCALE(data[0]);
+    ctx->color_encoding = TVG_HEADER_DATA_COLOR_ENC(data[0]);
+    ctx->coord_range = TVG_HEADER_DATA_RANGE(data[0]);
+    /* now read the width and height: */
+    uint32_t tmp;
+    tvg_result_t res = tvg_read_coord(ctx, &tmp);
+    if (res != TVG_SUCCESS) {
+        return res;
+    }
+    ctx->width = tvg_map_zero_to_max(ctx, tmp);
+    res = tvg_read_coord(ctx, &tmp);
+    if (res != TVG_SUCCESS) {
+        return res;
+    }
+    ctx->height = tvg_map_zero_to_max(ctx, tmp);
+    if (dim_only) {
+        return TVG_SUCCESS;
+    }
+    /* next read the color table */
+    uint32_t color_count;
+    res = tvg_read_varuint(ctx, &color_count);
+    if (res != TVG_SUCCESS) {
+        return res;
+    }
+    if (color_count == 0) {
+        return TVG_E_INVALID_FORMAT;
+    }
+    ctx->colors = malloc(color_count * sizeof(twin_argb32_t));
+    if (ctx->colors == NULL) {
+        return TVG_E_OUT_OF_MEMORY;
+    }
+    ctx->colors_size = (size_t) color_count;
+    for (size_t i = 0; i < ctx->colors_size; ++i) {
+        res = tvg_read_color(ctx, &ctx->colors[i]);
+        if (res != TVG_SUCCESS) {
+            free(ctx->colors);
+            ctx->colors = NULL;
+            return res;
+        }
+    }
+    return TVG_SUCCESS;
+}
+
+static tvg_result_t tvg_parse_gradient(tvg_context_t *ctx,
+                                       tvg_gradient_t *out_gradient)
+{
+    uint32_t u32;
+    tvg_point_t pt;
+    tvg_result_t res = tvg_read_point(ctx, &pt);
+    if (res != TVG_SUCCESS) {
+        return res;
+    }
+    out_gradient->point0 = pt;
+    res = tvg_read_point(ctx, &pt);
+    if (res != TVG_SUCCESS) {
+        return res;
+    }
+    out_gradient->point1 = pt;
+    res = tvg_read_varuint(ctx, &u32);
+    if (res != TVG_SUCCESS) {
+        return res;
+    }
+    out_gradient->color0 = u32;
+    if (u32 > ctx->colors_size) {
+        return TVG_E_INVALID_FORMAT;
+    }
+    res = tvg_read_varuint(ctx, &u32);
+    if (res != TVG_SUCCESS) {
+        return res;
+    }
+    if (u32 > ctx->colors_size) {
+        return TVG_E_INVALID_FORMAT;
+    }
+    out_gradient->color1 = u32;
+    return TVG_SUCCESS;
+}
+
+static tvg_result_t tvg_parse_style(tvg_context_t *ctx,
+                                    int kind,
+                                    tvg_style_t *out_style)
+{
+    tvg_result_t res;
```

Comment:
I cannot understand why `res` will be checked initially in line 557, but `res` will not be checked here?
```c
    uint32_t u32;
    tvg_result_t res = tvg_read_varuint(ctx, &u32);
    if (res != TVG_SUCCESS) {
        return res;
    }
```